rfc/performance-best-practices.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: Performance Best Practices
+- Status: draft
+- Start Date: 11-12-2017
+- Authors: Amruta-Ranade
+- Cockroach Doc Issue: n/a
+
+# Summary
+
+This document discusses the purpose and proposed documentation plan for documenting SQL performance best practices for CockroachDB.
+
+# Motivation
+
+Because CockroachDB is a distributed database, common database operations need to be done differently to achieve performance at par with traditional databases. Users might not be aware of these differences. Robert and support engineers already share these differences as best practices with customers on a one-on-one basis. However, they share it mostly after the customer hits performance issues and contacts us to resolve the issues. Documenting these best practices will help those customers who do not contact us, and will allow us to reach more customers. This may also help in increasing product adoption, education, and usage.
+
+# What are the recommended performance best practices?
+
+- Multi-row DML
+- Using `IMPORT TABLE` instead of `COPY`
+- Unique ID best practice: Use `RETURNING NOTHING` instead of transactions
+- Retrieving surrogate key to natural key
+- Transactions best practice: Use `RETURNING NOTHING`
+- JOINs: Merge `JOIN`, Hash `JOIN`
+- Secondary indexes
+- Performance implications of constraints: Have only as many constraints as you need. Example: One index vs. two indexes for `INSERT`/other DML statements would make it slower, but for `SELECT`, it would make it faster
+
+# What are other possible best practices?
+
+The following practices are recommended, but do not show significant performance improvements during performance tests:
+
+- Interleaved tables
+- Interleaved indexes
+- Column families
+
+The following practices are not yet supported for CockroachDB (but are on the roadmap):
+
+- Select * from non lease holders
+- CDC
+
+# Proposed documentation plan
+
+Add a note and example calling out multi-row DML best practices and using `RETURNING NOTHING` in the following existing docs:
+
+- `INSERT`
+- `UPSERT`
+- `DELETE`
+- `UPDATE`
+- `SELECT`
+- `TRANSACTIONS`
+
+Create new documents for:
+
+- Merge `JOIN`
+- Hash `JOIN`
+
+Create a Performance Best Practices document and link (or include) to the new documents listed above, as well as following existing documents:
```

Comment:
You also mentioned `Unique ID best practices` and `Retrieving surrogate key to natural key`. Where do those fit it?

Reply:
I added the content for `Unique ID best practices` in the relevant sections of the `INSERT` doc. As for `Retrieving surrogate key to natural key`, Robert's working on an example. Once I get that info, I can figure out which docs need to be updated and the relevant content.

Reply:
Sounds good.